modules/actions/snapshots/lib/apply-usd-values.ts

```diff
@@ -0,0 +1,88 @@
+import { Prisma, Chain } from '@prisma/client';
+import { prisma } from '../../../../prisma/prisma-client';
+import { now, roundToMidnight } from '../../../common/time';
+import _ from 'lodash';
+
+export const applyUSDValues = async (
+    rawSnapshots: Prisma.PrismaPoolSnapshotUncheckedCreateInput[],
+    fetchPrices: (chain: Chain, timestamp?: number) => Promise<Record<string, number>> = fetchPricesHelper,
+    fetchPoolTokens: (poolIds: string[]) => Promise<Record<string, any[]>> = fetchPoolTokensHelper,
+): Promise<Prisma.PrismaPoolSnapshotUncheckedCreateInput[]> => {
+    const lastMidnight = roundToMidnight(now());
+    const snapshots: Prisma.PrismaPoolSnapshotUncheckedCreateInput[] = [];
+
+    const poolIds = [...new Set(rawSnapshots.map((snapshot) => snapshot.poolId))];
+
+    // Pool tokens are needed, because SG returns raw token amounts
+    const poolTokens = await fetchPoolTokens(poolIds);
+
+    // For each timestamp, fetch the prices and calculate USD values
+    const groupedTimestamps = _.groupBy(rawSnapshots, 'timestamp');
+
+    for (const [timestamp, group] of Object.entries(groupedTimestamps)) {
+        // For each poolId, calculate USD values
+        const prices = await fetchPrices(
+            group[0].chain,
+            parseInt(timestamp) < lastMidnight ? parseInt(timestamp) : undefined,
+        );
+
+        for (const snapshot of group) {
+            if (
+                !poolTokens[snapshot.poolId] ||
+                poolTokens[snapshot.poolId].length === 0 ||
+                !snapshot.amounts ||
+                !snapshot.totalVolumes ||
+                !snapshot.totalProtocolSwapFees
+            ) {
+                snapshots.push(snapshot);
```

Comment:
what is this case, when does this happen?

Reply:
that's coming from SG